### PR TITLE
OS command injection sanitizer

### PR DESCRIPTION
**Pull request by bertschneider**
The OS command injection sanitizer checks if fuzzer provided data reaches command execution code in `ProcessImpl::start`.

Furthermore the PR includes an extension to the `@MethodHook` annotation. The `additionalClassesToHook` property can  provide an array of classes that should be included in the hook instrumentation.

### CODE REVIEW

File: agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java
Comment (fmeum) [inline agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java:23]: Could you also update the javadocs at https://github.com/CodeIntelligenceTesting/jazzer-api? I submitted a PR to get the instructions updated.

File: agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt
Comment (fmeum) [inline agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:60]: How about we add a FIXME comment here: With the current logic, every time we add a class to the new list because it was mentioned by a hook, _every_ custom hook will be applied to it. As we add more hooks, that may have undesired consequences, so at some point we might want to track the list of classes per custom hook rather than globally.

File: sanitizers/src/test/java/com/example/OsCommandInjectionProcessBuilder.java
Comment (fmeum) [inline sanitizers/src/test/java/com/example/OsCommandInjectionProcessBuilder.java:25]: Could you quickly verify what happens when this e.g. executes `shutdown` in the Bazel test sandbox? If that ends up shutting down the machine, we might need to take extra precautions here.

File: agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:145]: `,`  is deprecated in when-conditions. Another condition looked more cleanly to me but I don't have strong feeling about that.

File: agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java]: Changed.

File: agent/src/main/java/com/code_intelligence/jazzer/instrumentor/Hook.kt
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/instrumentor/Hook.kt]: After the default change it's not needed anymore. Changed it to `asList()` to retain type `List<String>` and be able to use the property in a `flatMap` later on.

File: agent/src/main/java/com/code_intelligence/jazzer/instrumentor/Hook.kt
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/instrumentor/Hook.kt]: It's of type `List<String>`, due to the `filter`,  but then again the `[ ]` are redundant. Removed them.

File: agent/src/main/java/com/code_intelligence/jazzer/utils/ClassNameGlobber.kt
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/utils/ClassNameGlobber.kt]: I don't have strong preferences either and think that was actually one of the older versions. Changed it.

File: sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt
Comment (bertschneider) [reply sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt:45]: The array size is checked in `ProcessBuilder.start` and an `IndexOutOfBoundsException` is thrown if it's empty. Didn't want to duplicate the check.

File: sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt
Comment (bertschneider) [reply sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt]: Changed.

File: sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt
Comment (bertschneider) [reply sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/OsCommandInjection.kt]: Changed.

File: agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java
Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/api/MethodHook.java:23]: Ah, was not aware the we have to do that. I will create a PR for the update.

File: sanitizers/src/test/java/com/example/OsCommandInjectionProcessBuilder.java
Comment (bertschneider) [reply sanitizers/src/test/java/com/example/OsCommandInjectionProcessBuilder.java:25]: At least no shutdown on my system :smile:

File: agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt
Comment (fmeum) [reply agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:145]: I didn't know that. https://kotlinlang.org/docs/control-flow.html#when-expression still contains examples of this. Do you know why this has been deprecated? I do prefer not having to repeat the body, even if it's simple.

**Review by fmeum — APPROVED** [review]

File: agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt
Comment (fmeum) [reply agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:145]: Weird, not sure why this got merged after my approval. The branch protection settings explicitly check for unresolved conversations... Well, we can just keep this as is, but please add any source for this deprecation you can find.

Comment (bertschneider) [reply agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:145]: It's handled in this ticket: https://youtrack.jetbrains.com/issue/KT-5143.
Oh, you merged the PR already...

Comment (fmeum) [reply agent/src/main/java/com/code_intelligence/jazzer/agent/RuntimeInstrumentor.kt:145]: I see, they recommend using `||` instead if the `when` has no argument. That makes sense and is what I should have proposed here. But as this has already been merged, let's just leave it as is.

**Pull request merged**
